cloud/pkg/taskmanager/nodeupgradecontroller/node_upgrade_controller.go

```diff
@@ -0,0 +1,351 @@
+		UpgradeID: upgrade.Name,
+		HistoryID: uuid.New().String(),
+		Version:   upgrade.Spec.Version,
+		Image:     image,
+	}
+
+	tolerate, err := strconv.ParseFloat(upgrade.Spec.FailureTolerate, 64)
+	if err != nil {
+		klog.Errorf("convert FailureTolerate to float64 failed: %v", err)
+		tolerate = 0.1
+	}
+
+	concurrency := upgrade.Spec.Concurrency
+	if concurrency <= 0 {
+		concurrency = 1
+	}
+	klog.V(4).Infof("deal task message: %v", upgrade)
+	ndc.MessageChan <- util.TaskMessage{
+		Type:            util.TaskUpgrade,
+		CheckItem:       upgrade.Spec.CheckItems,
+		Name:            upgrade.Name,
+		TimeOutSeconds:  upgrade.Spec.TimeoutSeconds,
+		Concurrency:     concurrency,
+		FailureTolerate: tolerate,
+		NodeNames:       upgrade.Spec.NodeNames,
+		LabelSelector:   upgrade.Spec.LabelSelector,
+		Status:          v1alpha1.TaskStatus{},
+		Msg:             upgradeReq,
+	}
+}
+
+func needUpgrade(node v1.Node, upgradeVersion string) bool {
+	if util.FilterVersion(node.Status.NodeInfo.KubeletVersion, upgradeVersion) {
+		klog.Warningf("Node(%s) version(%s) already on the expected version %s.", node.Name, node.Status.NodeInfo.KubeletVersion, upgradeVersion)
+		return false
+	}
+
+	// if node is in Upgrading state, don't need upgrade
+	if _, ok := node.Labels[util.NodeUpgradeJobStatusKey]; ok {
+		klog.Warningf("Node(%s) is in upgrade state", node.Name)
+		return false
+	}
+
+	return true
+}
+
+// nodeUpgradeJobDeleted is used to process deleted NodeUpgradeJob in apiserver
+func (ndc *NodeUpgradeController) nodeUpgradeJobDeleted(upgrade *v1alpha1.NodeUpgradeJob) {
+	// just need to delete from cache map
+	ndc.TaskManager.CacheMap.Delete(upgrade.Name)
+	klog.Errorf("upgrade job %s delete", upgrade.Name)
+	ndc.MessageChan <- util.TaskMessage{
+		Type:     util.TaskUpgrade,
+		Name:     upgrade.Name,
+		ShutDown: true,
+	}
+}
+
+// upgradeAdded is used to process update of new NodeUpgradeJob in apiserver
+func (ndc *NodeUpgradeController) nodeUpgradeJobUpdated(upgrade *v1alpha1.NodeUpgradeJob) {
+	oldValue, ok := ndc.TaskManager.CacheMap.Load(upgrade.Name)
+	old := oldValue.(*v1alpha1.NodeUpgradeJob)
+	if !ok {
+		klog.Infof("Upgrade %s not exist, and store it first", upgrade.Name)
+		// If Upgrade not present in Upgrade map means it is not modified and added.
+		ndc.nodeUpgradeJobAdded(upgrade)
+		return
+	}
+
+	// store in cache map
+	ndc.TaskManager.CacheMap.Store(upgrade.Name, upgrade)
+
+	node := checkUpdateNode(old, upgrade)
+	if node == nil {
+		klog.Info("none node update")
+		return
+	}
+
+	ndc.MessageChan <- util.TaskMessage{
+		Type:   util.TaskUpgrade,
+		Name:   upgrade.Name,
+		Status: *node,
+	}
+}
+
+func checkUpdateNode(old, new *v1alpha1.NodeUpgradeJob) *v1alpha1.TaskStatus {
+	if len(old.Status.Status) == 0 {
+		return nil
+	}
+	for i, updateNode := range new.Status.Status {
+		oldNode := old.Status.Status[i]
+		if !nodeUpdated(oldNode, updateNode) {
+			continue
+		}
+		return &updateNode
+	}
+	return nil
+}
```

Comment:
You can put a blank line here to make the code cleaner.

Reply:
done